main.cc

```diff
@@ -1343,6 +1343,7 @@ int main(int ac, char** av) {
 
             alternator::controller alternator_ctl(gossiper, proxy, mm, sys_dist_ks, cdc_generation_service, service_memory_limiter, *cfg);
             sharded<alternator::expiration_service> es;
+            std::any stop_expiration_service;
```

Comment:
Don't we already have an abort source we can use, instead of creating a new one? Stopping the main loop shouldn't be done in many different ways.

Reply:
This stop_expiration_service isn't an abort_source, it's a deferred call for actually calling the stop.
The expiration service has a separate abort_source per shard, I don't understand how it can work differently. Yes, we should have a global per-shard abort source, but if we do that, we would have no control over destruction order of the services...
Don't the other services work exactly like I did it - each service has its own stop function, which uses its own abort source (or whatever) to stop its own work?